### PR TITLE
chore(datastore): Enable cloud sync in integration tests

### PR DESCRIPTION
**Pull request by HuiSF**
*Issue #, if available:*

*Description of changes:*

- Forward native `subscritpionDataProcessed` (Android) and `syncReceived` (iOS) events to Dart DataStore hub event stream, these two events are named as `subscriptionDataProcessed` in dart implementation
- Updated test suites to run with cloud sync
- Split 23 testing models into smaller groups, and run `flutter test` separately to avoid long subscription and initial sync time which made tests run unstable in Android. 

NOTE: Android integration tests stability depends on PR #1339 ...
NOTE2: Cloud sync is enabled for model relationship tests and basic model operation within this PR. More tests with cloud sync will be add separately.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by dnys1 — COMMENTED** [review]
There is a lot of duplicated code in the tests - is there anyway to DRY this up a bit?

File: packages/amplify_datastore/example/integration_test/main_test.dart
Comment (HuiSF) [reply packages/amplify_datastore/example/integration_test/main_test.dart:22]: Hum meant to make it optional so can run in either mode, will modify the shell script a bit to make configurable.

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt
Comment (HuiSF) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt:54]: amplify-android sends nested models in `ModelWithMetadata` with meta fields (`_deleted`, `_version` and `_lastChangedAt`) merged into the model, and their values can be null, which may crash on `!!`.

File: build-support/integ_test_android.sh
Comment (ragingsquirrel3) [inline build-support/integ_test_android.sh:16]: question: Can the device argument be an optional argument passed to the script, with "sdk" as the default value? For some reason, in CI, "sdk" does not resolve and I've been passing "emulator-5554" instead, but I think it would help if this script accepted an argument for this.

File: packages/amplify_datastore/example/integration_test/utils/setup_utils.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/utils/setup_utils.dart:81]: Could this be changed to `await Amplify.DataStore.start().timeout(Duration(minutes: 2));` and remove the previous line and timeout util?

File: build-support/integ_test_android.sh
Comment (ragingsquirrel3) [inline build-support/integ_test_android.sh:28]: request: Could you please make this script such that it will have a non-zero exit status if any test command fails? I just tried forcing the first test suite to fail and and forcing the rest to pass, and even though output says "some tests failed" from failed test suite, melos thinks it passed, I think bc exit status of this script is still 0 if last suite it runs passed.

A good way to do this is to keep a list/array of failed tests, appending to it every time the `flutter test` command fails. Then, make the exit status the length of the list. no fails = status 0. You can see an example of that in test_all_plugins.sh

Same request for ios script.

File: packages/amplify_datastore/example/integration_test/separate_integration_tests/basic_model_operation_test.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/separate_integration_tests/basic_model_operation_test.dart:80]: question: Is it possible/desirable to test that it got the new name in addition to version bump?

File: packages/amplify_datastore/example/integration_test/separate_integration_tests/basic_model_operation_test.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/separate_integration_tests/basic_model_operation_test.dart:71]: question: Could there be some kind of race condition/error getting just the first blog from DB/API if some other client running the test? For API integ tests, uses a utility to create instances for each test. Maybe that's too much here, just asking. Same question for other tests.

While race condition might seem unlikely, always consider the CI will run android/ios in parallel.

File: packages/amplify_datastore/example/integration_test/separate_integration_tests/basic_model_operation_test.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/separate_integration_tests/basic_model_operation_test.dart:121]: question: Would it be better to try to query for the one you deleted in case race condition makes the backend not empty?

File: packages/amplify_datastore/example/integration_test/separate_integration_tests/has_one_explicit_child_test.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/separate_integration_tests/has_one_explicit_child_test.dart:60]: Would it be possible to have:
a) utilities for testing commonly repeated things like "precondition" "save child" "save parent" etc... as much of the code seems copy/pasted with small variations
b) a high-level utility, like `testCommonRelationshipOperations({bool saveChild = true, bool saveParent = true, ....})` that each of these tests could call, then just provide false for the operations they don't need to test

Higher-level goal to save some electricity/repetition here and make tests easier to maintain.

File: packages/amplify_datastore/example/integration_test/separate_integration_tests/multi_relationship_test.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/separate_integration_tests/multi_relationship_test.dart:194]: nit https://api.flutter.dev/flutter/package-matcher_matcher/containsAll.html, same for similar tests

File: packages/amplify_datastore/example/integration_test/utils/setup_utils.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/utils/setup_utils.dart:77]: Is this expected to be done before/after `Amplify.Datastore.start()`? Appears should be after? It's a little confusing to me, appears that we expect to start listening to hub, start datastore, then hub completes later.

File: packages/amplify_datastore/example/integration_test/utils/setup_utils.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/utils/setup_utils.dart:59]: nit: Does it make sense to make this delay a constant for the whole suite so it can be changed in one place? I believe the non-cloud tests do something similar IIRC.

File: packages/amplify_datastore/example/integration_test/utils/wait_for_expected_event_from_hub.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/utils/wait_for_expected_event_from_hub.dart:28]: nit: Same comment as in setup_utils.dart, but can we get rid of this utility and just call `.timeout` on the future in line 29?

File: packages/amplify_datastore/lib/amplify_datastore_stream_controller.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/lib/amplify_datastore_stream_controller.dart:46]: optional nit: I know it's already like this and a little tangential to PR, but can 'eventName' be made a constant? Ideally, lots of serialization keys would be in a constant somewhere,b ut this one is used several times right here, so could just be `const _eventName = 'eventName'` at top of file.

File: packages/amplify_datastore_plugin_interface/lib/src/types/hub/HubEventElement.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore_plugin_interface/lib/src/types/hub/HubEventElement.dart:22]: nice!

**Review by ragingsquirrel3 — CHANGES_REQUESTED** [review]
Nice job with this integ test change. It seems pretty stable for me running locally. There might be some challenges with this in CI, but not in scope for this PR IMO. A few high-level callouts:
1) Do we want the subscription events to be separate PR? If not, please reflect both changes in commit message when merged to it's there for anyone who is debugging/bisecting commits.
2) Like Dillon said, would be nice to DRY up the testing logic, left a suggestion for that, but feel free to chat offline more.
3) Test script should return non-zero status if anything fails.
4) Mostly nits after that.

**Comment by HuiSF** [comment]
> 1. Do we want the subscription events to be separate PR? If not, please reflect both changes in commit message when merged to it's there for anyone who is debugging/bisecting commits.

Could to a merge instead of squash merge on this PR (I'll squash comment resolving commits before merge).

File: packages/amplify_datastore/example/integration_test/utils/setup_utils.dart
Comment (HuiSF) [reply packages/amplify_datastore/example/integration_test/utils/setup_utils.dart:81]: The timeout is determined by how long it takes to receive the `ready` event from DataStore hub.
Using `timeout` to alter `Amplify.DataStore.start()` returned future is inaccurate I think...

Thinking of this point, `Amplify.DataStore.start()` is not need to be `await`'ed I'll remove it.

File: packages/amplify_datastore/example/integration_test/utils/setup_utils.dart
Comment (HuiSF) [reply packages/amplify_datastore/example/integration_test/utils/setup_utils.dart:77]: Hub gets ready in general for all possible events from any category.
So we can setup the listener to hub even before starting DataStore to ensure the even can be always captured.

In this implementation, `_completer.complete();` is invoked after `Amplify.DataStore.start()` completed. As the `start` API only starts storage engine and sync engine, as long as the starting completes, it returns, and not ensure if the ready event is dispatched or not.

Hope this explanation helps understanding.

File: packages/amplify_datastore/example/integration_test/separate_integration_tests/basic_model_operation_test.dart
Comment (HuiSF) [reply packages/amplify_datastore/example/integration_test/separate_integration_tests/basic_model_operation_test.dart:80]: There is no need to validate the updates on the name fields, as if the `_version` bumps, means AppSync + DynamoDB accepted the update mutation. The name field will certainly be updated.

File: build-support/integ_test_android.sh
Comment (HuiSF) [reply build-support/integ_test_android.sh]: Updated the script to allow passing flag arguments.

Supported arguments see [here](https://github.com/aws-amplify/amplify-flutter/pull/1351/files#diff-c2b60e08277abc187525d4594d19b5d56fe925f1e710af6f08bdd99b2f0eab28R13-R21).

File: build-support/integ_test_android.sh
Comment (HuiSF) [reply build-support/integ_test_android.sh]: I will look into this when integrating the integ tests runs into CI.

File: packages/amplify_datastore/example/integration_test/separate_integration_tests/has_one_explicit_child_test.dart
Comment (HuiSF) [reply packages/amplify_datastore/example/integration_test/separate_integration_tests/has_one_explicit_child_test.dart]: Added reusable test util function to simplify the scripts in each test suite.

File: packages/amplify_datastore/example/integration_test/main_test.dart
Comment (HuiSF) [reply packages/amplify_datastore/example/integration_test/main_test.dart:22]: Updated the melos run and shell script, that can run integration tests with cloud sync disabled.

```
melos run test:integration:android -- --enable-cloud-sync false
```

File: packages/amplify_datastore/example/integration_test/utils/setup_utils.dart
Comment (ragingsquirrel3) [reply packages/amplify_datastore/example/integration_test/utils/setup_utils.dart:79]: thanks

File: packages/amplify_datastore/example/integration_test/utils/setup_utils.dart
Comment (ragingsquirrel3) [inline packages/amplify_datastore/example/integration_test/utils/setup_utils.dart:73]: nit: can `async` be removed here as this returns future without calling `await`?

**Comment by ragingsquirrel3** [comment]
I think the datastore provision script needs a little fix to work with cloud sync, at https://github.com/aws-amplify/amplify-flutter/blob/main/packages/amplify_datastore/example/tool/add_api_request.json#L11 add:


```json
    "conflictResolution": {
      "defaultResolutionStrategy": {
        "type": "AUTOMERGE"
      }
    }
```

File: build-support/integ_test_ios.sh
Comment (ragingsquirrel3) [reply build-support/integ_test_ios.sh:72]: I think bc `set -euo pipefail` at top of script.

**Review by ragingsquirrel3 — APPROVED** [review]
A few small fixes needed and a nit, otherwise looks great! This improved so much, thanks for chipping away at a big, complicated problem and leaving us with this great script.

💯

File: build-support/integ_test_ios.sh
Comment (HuiSF) [reply build-support/integ_test_ios.sh:72]: Nice find! I will add the path check guard before running the script to run the special integration test suites.

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/DataStoreHubEventStreamHandler.kt
Comment (dnys1) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/DataStoreHubEventStreamHandler.kt:205]:
```suggestion
                        val eventData = hubEvent.data as ModelWithMetadata<*>
                        val model = eventData.model
                        if (model !is SerializedModel) {
                            LOG.error("Element is not an instance of SerializedModel.")
                            return
                        }
                        val message = FlutterSubscriptionDataProcessedEvent(
                            hubEvent.name,
                            eventData.syncMetadata,
                            model,
                        )
                        sendEvent(message.toValueMap())
```

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/hub/FlutterSubscriptionDataProcessedEvent.kt
Comment (dnys1) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/hub/FlutterSubscriptionDataProcessedEvent.kt:46]:
```suggestion
class FlutterSubscriptionDataProcessedEvent(
    override val eventName: String,
    private val syncMetadata: ModelMetadata,
    private val model: SerializedModel
) : FlutterHubEvent {
    override fun toValueMap(): Map<String, Any> {
        return mapOf(
            "eventName" to eventName,
            "modelName" to model.modelName,
            "element" to mapOf(
                "syncMetadata" to mapOf(
                    "id" to syncMetadata.id,
                    "_deleted" to isDeleted,
                    "_version" to syncMetadata.version,
                    "_lastChangedAt" to syncMetadata.lastChangedAt?.secondsSinceEpoch
                ),
                "model" to FlutterSerializedModel(model).toMap()
            )
        )
    }
}
```

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt
Comment (dnys1) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt:69]:
```suggestion
                            (value.cast<SerializedCustomType>()).map { item ->
```

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt
Comment (dnys1) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt:54]: Can you add that as a comment to the code?

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (dnys1) [inline packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:57]: Please use `guard` - force casts will crash the app if it ever fails

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (dnys1) [inline packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:64]:
```suggestion
            let serializedData = model["serializedData"] as? [String: Any] ?? [:]
```

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (dnys1) [inline packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:69]:
```suggestion
            self.deleted = serializedData["_deleted"] as? Bool
```

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterSyncReceivedEvent.swift
Comment (dnys1) [inline packages/amplify_datastore/ios/Classes/types/hub/FlutterSyncReceivedEvent.swift:45]:
```suggestion
            "element": self.element.toValueMap()
```

File: packages/amplify_datastore/lib/amplify_datastore_stream_controller.dart
Comment (dnys1) [inline packages/amplify_datastore/lib/amplify_datastore_stream_controller.dart:23]:
```dart
const channel = EventChannel('com.amazonaws.amplify/datastore_hub_events');
```

File: packages/amplify_datastore_plugin_interface/lib/src/types/hub/SubscriptionDataProcessedEvent.dart
Comment (dnys1) [inline packages/amplify_datastore_plugin_interface/lib/src/types/hub/SubscriptionDataProcessedEvent.dart:28]:
```suggestion
class SubscriptionDataProcessedEvent<M extends Model> extends HubEventPayload {
  final HubEventElementWithMetadata<M> element;
  final String modelName;

  SubscriptionDataProcessedEvent(this.element, this.modelName);

  factory SubscriptionDataProcessedEvent.fromMap(
    Map<dynamic, dynamic> serializedData,
    ModelProviderInterface provider,
    String eventName,
  ) {
    final element =
        HubEventElementWithMetadata<M>.fromMap(serializedData, provider);
    final modelName = serializedData["modelName"] as String;
    return SubscriptionDataProcessedEvent(element, modelName);
  }
}
```

**Review by dnys1 — APPROVED** [review]
Approved, contingent on changes mentioned

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (HuiSF) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:57]: Then it fails, the force cast exception will be captured and handled at this [spot](https://github.com/HuiSF/amplify-flutter/blob/chore/ds-integration-tests-cloud/packages/amplify_datastore/ios/Classes/DataStoreHubEventStreamHandler.swift#L190-L192).

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (HuiSF) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:64]: At this point, `model["serializedData"] as! [String: Any]` should always valid, as `try decodedModel.toMap()` should've completed without issues.

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (HuiSF) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:69]: Need to assign the default value `false` when `_delete` is `null`.
Will change to

```swift
let value = serializedData["_deleted"] as? Bool ?? false;
```

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (dnys1) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:57]: No, that's not true. It will crash the program - unwraps are not propagated as thrown errors like they are in Dart and other languages.

Screen Shot 2022-03-24 at 1 13 25 PM

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (dnys1) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:64]: You know that. I might even know that. But someone looking at this program for the first time doesn't know that, and the compiler doesn't know that. And those are the people that matter most.

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (dnys1) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:69]: Then make the property non-null as well on the type

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (HuiSF) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:57]: Good to know. But for this specific spot -

* `ModelRegistry.decode()` returns a `Model` otherwise it throws (would be captured)
* `FlutterSerializedModel` extends `Model`
* Within the DataStore iOS implementation, everything presented by `Model` is actually `FlutterSerializedModel`

Therefore this force unwrap should be safe here, Xcode hasn't prompt any warning as well.

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (HuiSF) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:69]: I can't, as `_deleted` is defined as nullable in a model schema.

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (HuiSF) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:64]: And this is a good point.

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (HuiSF) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:69]: Oh you mean the class property, yes I'll change it.

File: packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift
Comment (dnys1) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:57]: Like I said in my other comment, these are things you know but are not apparent by just reading the code. So please either write this in a comment above this line or just change it to use a `guard`/`throw` which, as you say, will have the same effect. The difference is that by changing it, it's ✨ **guaranteed** ✨ to never crash no matter what changes in our code or iOS.

You don't get a warning because you have statements that throw in the body. But your original statement was incorrect, this will crash the program if it ever fails. I'm trying to make a general point about force unwrapping, no longer specific to this one line.

Comment (dnys1) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:57]: In this case, you can also create a method with this comment and reuse that throughout our DataStore code, because it sounds like we may be doing this a lot throughout and there may be something we know that the compiler doesn't.

Comment (HuiSF) [reply packages/amplify_datastore/ios/Classes/types/hub/FlutterHubElement.swift:57]: Yea, the native impl. could use a good clean up. I'll fine some other time to do so... or waiting for the "final big" change. I'll take your suggestion here, thanks.

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/hub/FlutterSubscriptionDataProcessedEvent.kt
Comment (HuiSF) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/hub/FlutterSubscriptionDataProcessedEvent.kt:46]: Thanks, any advances for this proposed change on the constructor?
I kept the `eventName` and `modelName` as the first two parameter to make it similar with other event classes in the codebase.

Comment (dnys1) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/hub/FlutterSubscriptionDataProcessedEvent.kt:46]: Yes, there are no casts, and no duplicate data is being passed anymore.

Comment (dnys1) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/hub/FlutterSubscriptionDataProcessedEvent.kt:46]: It's okay to keep the other parameters for consistency, I was just trying to get things more concise.

File: packages/amplify_datastore_plugin_interface/lib/src/types/hub/SubscriptionDataProcessedEvent.dart
Comment (HuiSF) [reply packages/amplify_datastore_plugin_interface/lib/src/types/hub/SubscriptionDataProcessedEvent.dart:28]: I like this version better, but I will keep the current implementation following other event class implementations for now.

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt
Comment (HuiSF) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt]: This breaks for some cases when the value uses the `List` implementation is a variant other than `ArrayList`. E.g. `SingletonList` created by Kotlin `listOf`.

I think the `cast` function should keep the generic `List`?

Comment (dnys1) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt]: The definition of `cast` is incorrect... I'm not sure why because I had written it differently at one point. It should be

```kotlin
inline fun <reified T> List<*>.cast(): List<T> =
    this as List<T>
```

Comment (HuiSF) [reply packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterSerializedModel.kt]: I've updated the `cast` function with the latest commit, thanks.

**Pull request merged**
